types/resource.go

```diff
@@ -1,27 +1,309 @@
 package types
 
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+)
+
+type UnitType int
+
+var (
+	errCannotSub = errors.Errorf("cannot sub resources when lhs does not have same units as rhs")
+)
+
+type Attribute struct {
+	Key   string      `json:"key"`
+	Value interface{} `json:"value,omitempty"`
+}
+
+type Unit interface {
+	String() string
+	equals(Unit) bool
+	add(Unit)
+	sub(Unit)
+	le(Unit) bool
+}
+
+type ResUnit interface {
+	Equals(ResUnit) bool
+	Add(unit ResUnit) bool
+}
+
+// ResourceUnits describes all available resources types for deployment/node etc
+// if field is nil resource is not present in the given data-structure
+type ResourceUnits struct {
+	CPU     *CPU
+	Memory  *Memory
+	Storage *Storage
+}
+
+// Resources stores Unit details and Count value
+type Resources struct {
+	Units ResourceUnits `json:"units"`
+	Count uint32        `json:"count"`
+}
+
 // ResourceGroup is the interface that wraps GetName and GetResources methods
 type ResourceGroup interface {
 	GetName() string
-	GetResources() []Resource
+	GetResources() []Resources
+}
+
+type CPU struct {
+	Units      uint32                 `json:"units"`
+	Attributes map[string]interface{} `json:"attributes"`
+}
+
+type Memory struct {
+	Size       uint64                 `json:"size"`
+	Attributes map[string]interface{} `json:"attributes"`
+}
+
+type Storage struct {
+	Size       uint64                 `json:"size"`
+	Attributes map[string]interface{} `json:"attributes"`
+}
+
+var _ Unit = (*CPU)(nil)
+var _ Unit = (*Memory)(nil)
+var _ Unit = (*Storage)(nil)
+
+// AddUnit it rather searches for existing entry of the same type and sums values
+// if type not found it appends
+func (u ResourceUnits) Add(rhs ResourceUnits) (res ResourceUnits, err error) {
+	// check for in case some smartypants wants to overflow
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.Errorf("%v", r)
+		}
+
+		if err != nil {
+			res = ResourceUnits{}
+		}
+	}()
+
+	res = u
+
+	if res.CPU != nil {
+		res.CPU.add(rhs.CPU)
+	} else {
+		res.CPU = rhs.CPU
+	}
+
+	if res.Memory != nil {
+		res.Memory.add(rhs.Memory)
+	} else {
+		res.Memory = rhs.Memory
+	}
+
+	if res.Storage != nil {
+		res.Storage.add(rhs.Storage)
+	} else {
+		res.Storage = rhs.Storage
+	}
+
+	return
+}
+
+// Sub tbd
+func (u ResourceUnits) Sub(rhs ResourceUnits) (res ResourceUnits, err error) {
+	// check for in case some smartypants wants to overflow
+	defer func() {
+		if r := recover(); r != nil {
```

Comment:
let's just use big ints and return error if there's an overflow.  this catches too much.